### PR TITLE
Raise an error when trying to mark the current directory as vendored or data-only

### LINKED ISSUE #3019: Current directory as vendored is not working
## Actual Behavior

The `(vendored_dirs .)` stanza is not marking sub-directories as vendored.
I would expect this to work the same way as declaring the same directory as vendored from a parent directory:
`(vendored_dirs sub)` in `root/dune` vs
`(vendored_dirs .)` in `root/sub/dune`.

## Reproduction

For example, using `(vendored_dirs .)` instead of `(vendored_dirs *)` in `test/blackbox-tests/test-cases/vendor/duniverse/dune` changes the test output.

## Specifications

Using the dev version of Dune with OCaml `4.08.1`.

### PR DESCRIPTION
**Pull request by voodoos**
This PR add new error messages triggered when:
- Trying to declare the current directory as `data_only_dirs`
- Trying to declare the current directory as `vendored_dirs`

The check is quite naive and tricky globs like `../name-of-the-current-dir` will not trigger it.

Fixes #3019

### CODE REVIEW

File: src/dune/sub_dirs.ml [hunk 1]
```diff
@@ -170,8 +170,26 @@ let decode =
...
+        | None -> false
+      in
+      if is_dot data_only then
+        let loc = loc_or_not data_loc in
+        let hints = [ Pp.textf "did you mean %s?" "(data_only *)" ] in
+        User_error.raise ~loc ~hints
+          [ Pp.text "Only subdirectories can be marked as data_only." ]
+      else if is_dot vendored_dirs then
+        let loc = loc_or_not vendor_loc in
+        let hints = [ Pp.textf "did you mean %s?" "(vendored_dirs *)" ] in
```
Comment (rgrinberg) [inline src/dune/sub_dirs.ml]: Could you extract the generation of the hint into a helper function?

File: src/dune/sub_dirs.ml [hunk 2]
```diff
@@ -170,8 +170,26 @@ let decode =
...
-      let vendored_dirs = Option.map ~f:snd vendored_dirs in
-      (make ~dirs ~data_only ~ignored_sub_dirs ~vendored_dirs, rest)
+      let data_loc, data_only = Option.split data_only in
+      let vendor_loc, vendored_dirs = Option.split vendored_dirs in
+      let loc_or_not l = Option.value ~default:Loc.none l in
+      let is_dot pglob =
+        match pglob with
+        | Some pglob ->
+          Predicate_lang.Glob.exec pglob ~standard:Predicate_lang.empty "."
+        | None -> false
```
Comment (rgrinberg) [inline src/dune/sub_dirs.ml]: Small nit: we prefer to put the shorter clause on top.

**Review by rgrinberg — APPROVED** [review]
Looks good apart from some small nits. A change entry is appropriate as well.

**Comment by voodoos** [comment]
Do you think it would be worth it to have a smarter algorithm in the `is_dot` helper function line 176 ?

I was considering using `Build_system.eval_pred` to resolve the glob and then check if the current folder is in the list.

**Comment by rgrinberg** [comment]
> Do you think it would be worth it to have a smarter algorithm in the is_dot helper function line 176 ?

We could also disallow all path components if you'd like. In fact, I'm pretty sure we do this elsewhere when parsing globs.

> I was considering using Build_system.eval_pred to resolve the glob and then check if the current folder is in the list.

Is that possible? The glob should only resolve sub directories.

**Comment by voodoos** [comment]
Ok, the current PR is probably an acceptable compromise between complexity and helpfulness.

**Comment by rgrinberg** [comment]
> Ok, the current PR is probably an acceptable compromise between complexity and helpfulness.

Yeah, however what about sharing the validation that we have we for ignored sub dirs:

```
    let ignored =
      let+ l =
        enter
          (repeat
             (plain_string (fun ~loc dn ->
                  if
                    Filename.dirname dn <> Filename.current_dir_name
                    ||
                    match dn with
                    | ""
                    | "."
                    | ".." ->
                      true
                    | _ -> false
                  then
                    User_error.raise ~loc
                      [ Pp.textf "Invalid sub-directory name %S" dn ]
                  else
                    dn)))
      in
      Predicate_lang.Glob.of_string_set (String.Set.of_list l)
```

This prevents globs that are outside the current dir.

**Comment by voodoos** [comment]
I have the impression that this validation is too restrictive for `data_only_dirs` or `vendored_dirs`:

If I am not mistaken we want to allow:
- Direct subdirectories: `./dataonlydirs`
- Deeper subdirectories: `./somedir/dalaonlydir`

And disallow:
- Current directory: `.`
 
Should also be disallowed (but it may not be worth the trouble):
- Sneaky globs containing the current directory: `./somedir/../` or `../*`
- Or those pointing to a directory parent of the current directory: `..`

May also be allowed:
- Non-sub directories which are not parents of the current dir: `../dataonly`

**Comment by rgrinberg** [comment]
> Deeper subdirectories: ./somedir/dalaonlydir

We want to allow this but I don't think this currently works. We might as well reflect this in the validation.

> Should also be disallowed (but it may not be worth the trouble):

Agreed with disallowing all of this.

> May also be allowed:

I don't think we should allow this either. Directories shouldn't be able to affect how rules are setup in parent or sibling directories.

**Comment by voodoos** [comment]
> > Deeper subdirectories: ./somedir/dalaonlydir
> 
> We want to allow this but I don't think this currently works. We might as well reflect this in the validation.

I just checked and indeed it doesn't work. It should indeed be reflected by the validation, I will start reworking the PR.

**Comment by rgrinberg** [comment]
Would be good to have an explicit test for this as well. It’s something that we’d like to fix in the future of course, so feel free on investigating what it would take to extend this functionality

**Comment by voodoos** [comment]
> Yeah, however what about sharing the validation that we have we for ignored sub dirs:
> 
> ```
>    ...
>       Predicate_lang.Glob.of_string_set (String.Set.of_list l)
> ```
> 
> This prevents globs that are outside the current dir.

It also prevent `*` to function, in fact, if I understand well, the same result could be obtained with a simple string. Thus I cannot use it for `vendored_dirs` or `data_only_dirs`  because we want to be able to vendor all subfolders with `*`.

I am unsure of how I could adapt the present validation to work as intended with `*`.

**Comment by voodoos** [comment]
I guess I can use something like that instead:

`Predicate_lang.Glob.of_glob (Glob.of_string_exn loc some_string)`

Will give it a try.

File: test/blackbox-tests/test-cases/vendor/run.t [hunk 3]
```diff
@@ -40,3 +40,45 @@ The same directory cannot be marked as both vendored and data-only
...
+  [1]
+
+Only direct subdirectories can be marked as data-only
+
+  $ dune build --root deep-subfolder-dataonly
+  Entering directory 'deep-subfolder-dataonly'
+  File "dune", line 1, characters 16-19:
+  1 | (data_only_dirs a/b)
+                      ^^^
+  Error: Invalid sub-directory name "a/b"
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/vendor/run.t]: I think the error message can be improved in this case. Something like:

```
Error: Only immediate sub-directories may be specified. To ignore "a/b", write "(data_only_dirs b)" in a/dune
```

Comment (voodoos) [reply test/blackbox-tests/test-cases/vendor/run.t]: Good idea, it should work with 3b1be6c

**Pull request merged**